src/backend/metal/MetalBackend.mm

```diff
@@ -179,685 +200,67 @@ void Present(nxtDevice device) {
-                    // TODO(cwallez@chromium.org): does any state need to be reset?
-                }
-            }
-            void BeginSubpass(id<MTLCommandBuffer> commandBuffer, uint32_t subpass) {
-                ASSERT(currentRenderPass);
-                if (render != nil) {
-                    [render endEncoding];
-                    render = nil;
-                }
-
-                const auto& info = currentRenderPass->GetSubpassInfo(subpass);
-
-                MTLRenderPassDescriptor* descriptor = [MTLRenderPassDescriptor renderPassDescriptor];
-                bool usingBackbuffer = false; // HACK(kainino@chromium.org): workaround for not having depth attachments
-                for (uint32_t index = 0; index < info.colorAttachments.size(); ++index) {
-                    uint32_t attachment = info.colorAttachments[index];
-
-                    // TODO(kainino@chromium.org): currently a 'null' texture view
-                    // falls back to the 'back buffer' but this should go away
-                    // when we have WSI.
-                    id<MTLTexture> texture = nil;
-                    if (auto textureView = currentFramebuffer->GetTextureView(attachment)) {
-                        texture = ToBackend(textureView->GetTexture())->GetMTLTexture();
-                    } else {
-                        texture = device->GetCurrentTexture();
-                        usingBackbuffer = true;
-                    }
-                    descriptor.colorAttachments[index].texture = texture;
-                    descriptor.colorAttachments[index].loadAction = MTLLoadActionLoad;
-                    descriptor.colorAttachments[index].storeAction = MTLStoreActionStore;
-                }
-                // TODO(kainino@chromium.org): load depth attachment from subpass
-                if (usingBackbuffer) {
-                    descriptor.depthAttachment.texture = device->GetCurrentDepthTexture();
-                    descriptor.depthAttachment.loadAction = MTLLoadActionLoad;
-                    descriptor.depthAttachment.storeAction = MTLStoreActionStore;
-                }
-
-                render = [commandBuffer renderCommandEncoderWithDescriptor:descriptor];
-                // TODO(cwallez@chromium.org): does any state need to be reset?
-            }
-            void EndRenderPass() {
-                ASSERT(render != nil);
-                [render endEncoding];
-                render = nil;
-            }
-        };
-
+    uint64_t Device::GetPendingCommandSerial() {
+        // If this is called, then it means some piece of code somewhere will wait for this serial to
+        // complete. Make sure the pending command buffer is created so that it is on the worst case
+        // enqueued on the next Tick() and eventually increments the serial. Otherwise if no GPU work
+        // happens we could be waiting for this serial forever.
+        GetPendingCommandBuffer();
```

Comment:
The problem was that the end2end tests defer expectations and mapping the buffers for reading until the last graphic commands have been enqueued. Mapping buffers doesn't use the pending command buffer, but still needs to know when the GPU is done executing the current commands.

The idea is that if anything asks what it the pending command serial, then it will certainly watch for it to be finished, and even if we don't have any work to send to the GPU, we should still put the signal to update the serial.